File: app/src/main/kotlin/com/learner/lm/ai/SubjectClassifier.kt
package com.learner.lm.ai

class SubjectClassifier {

    private val mathKeywords = setOf(
        "equation", "solve", "algebra", "geometry", "fraction", "percent",
        "multiply", "divide", "add", "subtract", "variable", "graph", "theorem"
    )
    private val scienceKeywords = setOf(
        "atom", "molecule", "cell", "photosynthesis", "gravity", "force",
        "energy", "chemical", "biology", "physics", "experiment", "hypothesis"
    )
    private val englishKeywords = setOf(
        "essay", "grammar", "verb", "noun", "paragraph", "thesis", "literature",
        "poem", "writing", "sentence", "punctuation", "analyze", "theme"
    )
    private val historyKeywords = setOf(
        "war", "revolution", "empire", "century", "president", "civilization",
        "timeline", "ancient", "colony", "treaty", "dynasty"
    )
    private val geographyKeywords = setOf(
        "continent", "country", "climate", "map", "latitude", "longitude",
        "river", "mountain", "population", "region", "ecosystem"
    )

    fun classify(text: String): Subject {
        val normalized = text.lowercase()
        val scores = mapOf(
            Subject.MATH to countMatches(normalized, mathKeywords),
            Subject.SCIENCE to countMatches(normalized, scienceKeywords),
            Subject.ENGLISH to countMatches(normalized, englishKeywords),
            Subject.HISTORY to countMatches(normalized, historyKeywords),
            Subject.GEOGRAPHY to countMatches(normalized, geographyKeywords)
        )
        val best = scores.maxByOrNull { it.value }
        return if (best == null || best.value == 0) Subject.GENERAL else best.key
    }

    private fun countMatches(text: String, keywords: Set<String>): Int =
        keywords.count { keyword -> text.contains(keyword) }
}

File: workers/src/learner_api/services/subject_classifier.py
from learner_api.schemas import BuiltinSubject


class SubjectClassifier:
  MATH_KEYWORDS = {
      "equation", "solve", "algebra", "geometry", "fraction", "percent",
      "multiply", "divide", "add", "subtract", "variable", "graph", "theorem",
  }
  SCIENCE_KEYWORDS = {
      "atom", "molecule", "cell", "photosynthesis", "gravity", "force",
      "energy", "chemical", "biology", "physics", "experiment", "hypothesis",
  }
  ENGLISH_KEYWORDS = {
      "essay", "grammar", "verb", "noun", "paragraph", "thesis", "literature",
      "poem", "writing", "sentence", "punctuation", "analyze", "theme",
  }
  HISTORY_KEYWORDS = {
      "war", "revolution", "empire", "century", "president", "civilization",
      "timeline", "ancient", "colony", "treaty", "dynasty",
  }
  GEOGRAPHY_KEYWORDS = {
      "continent", "country", "climate", "map", "latitude", "longitude",
      "river", "mountain", "population", "region", "ecosystem",
  }

  def classify(self, text: str) -> BuiltinSubject:
      normalized = text.lower()
      scores = {
          BuiltinSubject.MATH: self._count_matches(normalized, self.MATH_KEYWORDS),
          BuiltinSubject.SCIENCE: self._count_matches(normalized, self.SCIENCE_KEYWORDS),
          BuiltinSubject.ENGLISH: self._count_matches(normalized, self.ENGLISH_KEYWORDS),
          BuiltinSubject.HISTORY: self._count_matches(normalized, self.HISTORY_KEYWORDS),
          BuiltinSubject.GEOGRAPHY: self._count_matches(normalized, self.GEOGRAPHY_KEYWORDS),
      }
      best = max(scores, key=scores.get)
      return best if scores[best] > 0 else BuiltinSubject.GENERAL

  @staticmethod
  def _count_matches(text: str, keywords: set[str]) -> int:
      return sum(1 for keyword in keywords if keyword in text)
